crates/ty_python_semantic/src/types/infer.rs

```diff
@@ -405,55 +413,130 @@ struct TypeAndRange<'db> {
     range: TextRange,
 }
 
-/// The inferred types for a single region.
+/// The inferred types for a scope region.
 #[derive(Debug, Eq, PartialEq, salsa::Update, get_size2::GetSize)]
-pub(crate) struct TypeInference<'db> {
+pub(crate) struct ScopeInference<'db> {
```

Comment:
Wondering if you also considered making `TypeInference` and `TypeInferenceBuilder` generic over the inference region kind (scope, expression, definition)? This might avoid some of the code duplication, but could be a bit tricky/annoying to set up without enum support in const generics.

Reply:
I wanted to avoid making `TypeInferenceBuilder` generic because it would lead to a lot of monomorphization and also just complicates the type signature overall. 

My first version had a `TypeInference` enum but I then realized that it's actually never needed (and we would still need to unwrap at the caller side to get the narrower type that we can put into the queries. But maybe you had something else in mind?

Reply:
> I wanted to avoid making `TypeInferenceBuilder` generic because it would lead to a lot of monomorphization and also just complicates the type signature overall.

Thanks. I figured you already thought about it.

> My first version had a `TypeInference` enum but I then realized that it's actually never needed (and we would still need to unwrap at the caller side to get the narrower type that we can put into the queries. But maybe you had something else in mind?

I only said enum because I would be inclined to make these structs generic over a `{ Scope, Expression, Definition }` enum. That works in C++, but not (yet?) with Rust const generics. Anyway, thanks for your response.